### PR TITLE
Drop zxcvbn password dependency

### LINKED ISSUE #121: Reconsider dependency to zxcvbn due to its size
Looking for something in the built file `formol.js` I noticed a huge word list.

image

This is coming from our dependency [zxcvbn](https://github.com/dropbox/zxcvbn) to estimate password strength.

While there are reasons behind it:
* https://github.com/dropbox/zxcvbn/issues/169
* https://www.usenix.org/conference/usenixsecurity16/technical-sessions/presentation/wheeler
* https://github.com/dropbox/zxcvbn/issues/59

We might reconsider this dependency as its size is 823 kB, upon a total of 1.5 MB of the `formol.js` file.

### PR DESCRIPTION
**Pull request by alexandreNoens**
zxcvbn adds a large embedded dictionary to formol.js. For issue #121, this dependency represented about 823 kB and a significant part of the client bundle.

Replace zxcvbn scoring in PasswordStrengthField with an internal password-strength utility based on:
- length bounds (min and max)
- repeated character detection
- sequential and keyboard pattern detection
- email-like input detection
- entropy thresholds mapped to scores 0..4

Update defaults to minLength=8 and maxLength=64, remove userInputs from the scoring path, and add focused tests for both the new helper and the password-strength field behavior.

Side effects: password scores can differ from previous zxcvbn results, especially for edge cases, but validation flow is preserved while reducing bundle size.

Close #121

### CODE REVIEW

File: src/utils/password.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    return 1
+  }
+  if (entropy < ENTROPY_THRESHOLDS.good) {
+    return 2
+  }
+  if (entropy < ENTROPY_THRESHOLDS.strong) {
+    return 3
+  }
+  return 4
+}
```
Comment (glepretre) [inline src/utils/password.js:107]: Could be shortened I think

File: src/utils/password.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    return { score: 1 }
+  }
+
+  if (hasSequentialPattern(value)) {
+    return { score: 1 }
+  }
+
+  if (isEmailAddress(value)) {
+    return { score: 2 }
+  }
```
Comment (glepretre) [inline src/utils/password.js]: Let's factorize and always return 1 for all anti-patterns

File: test/utils/password.test.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        4,
+        '64 chars accepted',
+      ],
+      [
+        'bbNBFt2okt4H03fdrFqlwigHxakPPdCZMEbt65MesByuBC1mxaoHRiYwab4FKjLq5',
+        8,
+        64,
+        0,
+        '65 chars rejected',
+      ],
```
Comment (glepretre) [inline test/utils/password.test.js]: IRL prevented by html maxLength attribute but good test

File: src/utils/password.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  return parts
+}
+
+function entropyOfSegment(seg) {
+  switch (seg.type) {
+    case 'digits':
+      return Math.log2(10) * seg.value.join('').length
+
+    case 'symbols':
+      return Math.log2(30) * seg.value.join('').length
```
Comment (glepretre) [inline src/utils/password.js]: 30 might be a bit low as symbol alphabet length

File: src/utils/password.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    return { score: 1 }
+  }
+
+  if (isEmailAddress(value)) {
+    return { score: 2 }
+  }
+
+  const entropy = calculateTotalEntropy(value)
+  const entropyScore = entropyToScore(entropy)
+  return { score: entropyScore }
```
Comment (glepretre) [inline src/utils/password.js]: We should have a lengthScore too and return lowest of (length, entropy)

File: src/utils/password.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+function entropyOfSegment(seg) {
+  switch (seg.type) {
+    case 'digits':
+      return Math.log2(10) * seg.value.join('').length
+
+    case 'symbols':
+      return Math.log2(30) * seg.value.join('').length
+
+    case 'letters':
+      return Math.log2(26) * seg.value.join('').length
```
Comment (glepretre) [inline src/utils/password.js]: 26 is only 1 case of latin alphabet without any accent.

File: src/utils/password.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+  const digits = str.match(/[0-9]+/g)
+  if (digits) {
+    parts.push({ type: 'digits', value: digits })
+  }
+
+  const symbols = str.match(/[^a-zA-Z0-9]+/g)
+  if (symbols) {
+    parts.push({ type: 'symbols', value: symbols })
+  }
```
Comment (glepretre) [inline src/utils/password.js]: we should check french accents and more unicode support and impact on score

File: src/utils/password.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
+const SEQUENTIAL_PATTERNS = [
+  /(0123|1234|2345|3456|4567|5678|6789|7890)/,
+  /(0987|9876|8765|7654|6543|5432|4321|3210)/,
+  /(abcd|bcde|cdef|defg|efgh|fghi|ghij|hijk|ijkl|jklm|klmn|lmno|mnop|nopq|opqr|pqrs|qrst|rstu|stuv|tuvw|uvwx|vwxy|wxyz)/i,
+  /(zyxw|yxwv|xwvu|wvut|vuts|utsr|tsrq|srqp|rqpo|qpon|ponm|onml|nmlk|mlkj|lkji|kjih|jihg|ihgf|hgfe|gfed|fedc|edcb|dcba)/i,
+  /(azer|zert|erty|qsdf|sdfg|wxcv|xcvb)/i,
+  /(qwer|wert|erty|asdf|sdfg|zxcv|xcvb)/i,
+]
```
Comment (glepretre) [inline src/utils/password.js]: Can be shortened I think

**Review by glepretre — CHANGES_REQUESTED** [review]
Very good start, we may now use AI to strengthen it 💪 You can handle the issues on Wednesday and we'll finish in pair-programming on Thursday @alexandreNoens

File: src/utils/password.js (hunk 8 above)
Comment (gentooboontoo) [reply src/utils/password.js]: Or dynamically generated at first use.

File: src/utils/password.js (hunk 6 above)
Comment (gentooboontoo) [reply src/utils/password.js]: Does not take into account case too (uppercase / lowercase)

**Review by gentooboontoo — COMMENTED** [review]
Some feedback but 👍

File: src/utils/password.js (hunk 2 above)
Comment (alexandreNoens) [reply src/utils/password.js]: I don't think we should collapse all antipatterns to score 1.

Repeated characters and obvious sequential patterns are strong signals of a weak password, so returning 1 makes sense there :+1: 

An email-like substring is different: it can also appear incidentally inside an otherwise strong password because of the `@` separator. In that case, forcing the score down to `1` would over-penalize the password and classify it as weak, even though the email match was not the dominant pattern.

That is why I kept `isEmailAddress(value)` at score `2`: it still penalizes the password, but less aggressively than patterns that are much more clearly weak by construction.

Comment (glepretre) [reply src/utils/password.js]: Understood, I agree

**Review by glepretre — APPROVED** [review]
@alexandreNoens The full diff is good 👏🎉  Feel free to rebase and squash commits, you can merge all in one or keep atomic commits you find relevant. We'll release this on Monday morning 🙂 

GG 💪

**Comment by glepretre** [comment]
## formol.js bundle size analysis :

| master | PR |
|--------|--------|
| 1.5 MB | 728 kB |

 zxcvbn lib size : 823 kB
 formol.js difference: -817 kB 
 
  :weight_lifting: :arrow_right: :feather: 
  
Behavior preserved and tested :ok: 
Mission complete :heavy_check_mark:
